### PR TITLE
gh-121849: Fix PyUnicodeWriter_WriteSubstring() crash if len=0

### PR DESCRIPTION
**Pull request by vstinner**
Do nothing if start=end.

* Issue: gh-121849

### CODE REVIEW

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -13637,27 +13637,28 @@ int
...
                                 Py_ssize_t start, Py_ssize_t end)
 {
-    Py_UCS4 maxchar;
-    Py_ssize_t len;
-
     assert(0 <= start);
     assert(end <= PyUnicode_GET_LENGTH(str));
     assert(start <= end);
 
-    if (end == 0)
```
Comment (picnixz) [inline Objects/unicodeobject.c:13647]: Funny that we only considered `start = end = 0` and not `start = end`.

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -13637,27 +13637,28 @@ int
...
-        return 0;
-
     if (start == 0 && end == PyUnicode_GET_LENGTH(str))
         return _PyUnicodeWriter_WriteStr(writer, str);
 
-    if (PyUnicode_MAX_CHAR_VALUE(str) > writer->maxchar)
+    Py_ssize_t len = end - start;
```
Comment (picnixz) [inline Objects/unicodeobject.c:13647]: EDIT: NVM my suggestion.

Comment (picnixz) [reply Objects/unicodeobject.c:13647]: Wait. My logic is flawed.

**Review by picnixz — APPROVED** [review]
Forget my previous comment. I was still sleepy.

File: Objects/unicodeobject.c (hunk 1 above)
Comment (vstinner) [reply Objects/unicodeobject.c:13647]: (end-start)=0 should cover all cases.

**Pull request merged**